lib/rosette/integrations/smartling-integration/smartling_completer.rb

```diff
@@ -1,49 +1,164 @@
 # encoding: UTF-8
 
+require 'thread'
+require 'concurrent'
+
 module Rosette
   module Integrations
     class SmartlingIntegration < Integration
       class SmartlingCompleter
 
-        attr_reader :configuration, :smartling_apis, :pullers
+        DEFAULT_THREAD_POOL_SIZE = 10
+
+        attr_reader :rosette_config, :repo_config
+        attr_reader :thread_pool_size
+
+        def initialize(rosette_config)
+          @rosette_config = rosette_config
+          @thread_pool_size = DEFAULT_THREAD_POOL_SIZE
+        end
 
-        def initialize(configuration, smartling_apis, pullers)
-          @smartling_apis = smartling_apis
-          @configuration = configuration
-          @pullers = pullers
+        def set_repo_config(repo_config)
+          @repo_config = repo_config
+          self
         end
 
-        def complete(repo_name, locales)
-          file_hash ||= Hash.new { |h, key| h[key] = [] }
-          file_uris_to_files = {}
+        def set_thread_pool_size(size)
+          @thread_pool_size = size
+          self
+        end
 
-          locales.each do |locale|
-            puller = puller_for_locale(locale)
-            completed_files = puller.completed_files_map[repo_name]
-            completed_files.each do |file|
-              file_uris_to_files[file.file_uri] ||= file
-              file_hash[file.file_uri] << locale
+        def complete
+          build_completion_maps.each do |file_uri, locales|
+            if locales.all? { |_, v| v.complete? }
+              begin
+                file = locales[repo_config.locales.first.code]
```

Comment:
nitpick: without validation, the repo config could not have any locales and cause an error here.